File: test/language/parameterized_types/java/parameterized_types/grand_child_param/GrandChildParamTest.java
package parameterized_types.grand_child_param;

import static org.junit.Assert.assertEquals;

import java.io.File;
import java.io.IOException;

import org.junit.Test;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.FileBitStreamReader;

public class GrandChildParamTest
{
    @Test
    public void fileWrite() throws IOException
    {
        final GrandChildParam grandChildParam = createGrandChildParam();
        final File file = new File(BLOB_NAME);
        grandChildParam.write(file);
        final BitStreamReader reader = new FileBitStreamReader(file);
        checkGrandChildParamInStream(reader, grandChildParam);
        reader.close();
        final GrandChildParam readGrandChildParam = new GrandChildParam(file);
        assertEquals(grandChildParam, readGrandChildParam);
    }

    private ItemChoiceHolder createItemChoiceHolder()
    {
        final Item item = new Item(ITEM_CHOICE_HOLDER_HAS_ITEM, ITEM_PARAM, ITEM_EXTRA_PARAM);
        ItemChoice itemChoice = new ItemChoice(ITEM_CHOICE_HOLDER_HAS_ITEM);
        itemChoice.setItem(item);

        return new ItemChoiceHolder(itemChoice.getHasItem(), itemChoice);
    }

    private GrandChildParam createGrandChildParam()
    {
        final ItemChoiceHolder itemChoiceHolder = createItemChoiceHolder();
        final ItemChoiceHolder[] itemChoiceHolderArray = new ItemChoiceHolder[] {createItemChoiceHolder()};
        final long[] dummyArray = new long[] {0};

        return new GrandChildParam(itemChoiceHolder, itemChoiceHolderArray, dummyArray);
    }

    private void checkItemChoiceHolderInStream(BitStreamReader reader, ItemChoiceHolder itemChoiceHolder)
            throws IOException
    {
        assertEquals(itemChoiceHolder.getHasItem(), reader.readBool());

        final Item item = itemChoiceHolder.getItemChoice().getItem();
        assertEquals(item.getParam(), reader.readBits(16));
        assertEquals((long)item.getExtraParam(), reader.readBits(32));
    }

    private void checkGrandChildParamInStream(BitStreamReader reader, GrandChildParam grandChildParam)
            throws IOException
    {
        final ItemChoiceHolder itemChoiceHolder = grandChildParam.getItemChoiceHolder();
        checkItemChoiceHolderInStream(reader, itemChoiceHolder);

        final ItemChoiceHolder[] itemChoiceHolderArray = grandChildParam.getItemChoiceHolderArray();
        assertEquals(itemChoiceHolderArray.length, reader.readVarSize());
        checkItemChoiceHolderInStream(reader, itemChoiceHolderArray[0]);

        final boolean isDummyArrayUsed = grandChildParam.isDummyArrayUsed();
        assertEquals(isDummyArrayUsed, reader.readBool());
        if (isDummyArrayUsed)
        {
            final long[] dummyArray = grandChildParam.getDummyArray();
            assertEquals(dummyArray.length, reader.readVarSize());
            assertEquals(dummyArray[0], reader.readBits(32));
        }
    }

    private static final String BLOB_NAME = "grand_child_param.blob";
    private static final boolean ITEM_CHOICE_HOLDER_HAS_ITEM = true;
    private static final int ITEM_PARAM = 0xAABB;
    private static final long ITEM_EXTRA_PARAM = 0x11223344;
}

File: test/language/parameterized_types/cpp98/GrandChildParamTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "parameterized_types/grand_child_param/GrandChildParam.h"

namespace parameterized_types
{
namespace grand_child_param
{

class GrandChildParamTest : public ::testing::Test
{
protected:
    void fillGrandChildParam(GrandChildParam& grandChildParam)
    {
        ItemChoiceHolder& itemChoiceHolder = grandChildParam.getItemChoiceHolder();
        itemChoiceHolder.setHasItem(ITEM_CHOICE_HOLDER_HAS_ITEM);
        ItemChoice& itemChoice = itemChoiceHolder.getItemChoice();
        Item item;
        item.setParam(ITEM_PARAM);
        item.setExtraParam(ITEM_EXTRA_PARAM);
        itemChoice.setItem(item);
    }

    void checkGrandChildParamInBitStream(zserio::BitStreamReader& reader,
            const GrandChildParam& grandChildParam)
    {
        const ItemChoiceHolder& itemChoiceHolder = grandChildParam.getItemChoiceHolder();

        ASSERT_EQ(itemChoiceHolder.getHasItem(), reader.readBool());

        const Item& item = itemChoiceHolder.getItemChoice().getItem();
        ASSERT_EQ(item.getParam(), reader.readBits(16));
        ASSERT_EQ(item.getExtraParam(), reader.readBits(32));
    }

private:
    static const bool     ITEM_CHOICE_HOLDER_HAS_ITEM;
    static const uint16_t ITEM_PARAM;
    static const uint32_t ITEM_EXTRA_PARAM;
};

const bool     GrandChildParamTest::ITEM_CHOICE_HOLDER_HAS_ITEM = true;
const uint16_t GrandChildParamTest::ITEM_PARAM = 0xAABB;
const uint32_t GrandChildParamTest::ITEM_EXTRA_PARAM = 0x11223344;

TEST_F(GrandChildParamTest, write)
{
    GrandChildParam grandChildParam;
    fillGrandChildParam(grandChildParam);

    zserio::BitStreamWriter writer;
    grandChildParam.write(writer);

    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    checkGrandChildParamInBitStream(reader, grandChildParam);
    reader.setBitPosition(0);

    GrandChildParam readGrandChildParam(reader);
    ASSERT_EQ(grandChildParam, readGrandChildParam);
}

} // namespace grand_child_param
} // namespace parameterized_types
